### PR TITLE
[Merged by Bors] - Gltf animations

### PR DESCRIPTION
**Pull request by mockersf**
# Objective

- Load informations for animations from GLTF
- Make experimenting on animations easier

# Non Objective

- Implement a solutions for all animations in Bevy. This would need a discussion / RFC. The goal here is only to have the information available to try different APIs

## Solution

- Load animations with a representation close to the GLTF spec
- Add an example to display animations. There is an animation driver in the example, not in Bevy code, to show how it can be used. The example is cycling between examples from the official gltf sample ([AnimatedTriangle](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/AnimatedTriangle), [BoxAnimated](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/BoxAnimated)), and one from me with some cases not present in the official examples.

https://user-images.githubusercontent.com/8672791/150696656-073403f0-d921-43b6-beaf-099c7aee16ed.mp4

### CODE REVIEW

File: examples/3d/load_gltf_animation.rs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+                start_time: time.seconds_since_startup(),
+                animation: animation.clone(),
+            })
+        }
+    }
+}
+
+// This animation driver is not made to work in the general case. It will work with only one
+// animation per scene, and will ignore the specified interpolation method to only do linear.
+fn gltf_animation_driver(
```
Comment (StarArawn) [inline examples/3d/load_gltf_animation.rs:150]: It's probably worth pointing out here that this isn't completely implemented per the spec. It in fact only animates transforms for GLTF nodes and will not animated bones or morphs which are also apart of the spec. I don't see any changes to the GLTF loader to load the skin information as well. I think this is a great PR though and I don't think those things should hold back it! We should just make it clear what's not currently supported here. 🤔

Comment (mockersf) [reply examples/3d/load_gltf_animation.rs:150]: Skin will be added in #2359, and morphs in #3722. This PR should work directly for skin, but morph will require additional work to be animated, currently they will log a warning when loading a file with morph animation

File: crates/bevy_gltf/src/loader.rs [hunk 2]
```diff
@@ -109,6 +112,76 @@ async fn load_gltf<'a, 'b>(
...
+                    gltf::animation::util::ReadOutputs::Scales(scale) => {
+                        GltfNodeAnimationKeyframes::Scale(scale.map(Vec3::from).collect())
+                    }
+                    gltf::animation::util::ReadOutputs::MorphTargetWeights(_) => {
+                        warn!("Morph animation property not yet supported");
+                        continue;
+                    }
+                }
+            } else {
+                panic!("animations without a sampler output are not supported");
```
Comment (superdump) [inline crates/bevy_gltf/src/loader.rs:161]: Same.

File: crates/bevy_gltf/src/lib.rs [hunk 3]
```diff
@@ -63,3 +69,52 @@ pub struct GltfPrimitive {
     pub mesh: Handle<Mesh>,
     pub material: Option<Handle<StandardMaterial>>,
 }
+
+/// Part of a [`GltfNodeAnimation`], interpolation method for an animation.
```
Comment (superdump) [inline crates/bevy_gltf/src/lib.rs:73]: First what it is, then context.
```suggestion
/// Interpolation method for an animation. Part of a [`GltfNodeAnimation`].
```

File: crates/bevy_gltf/src/lib.rs [hunk 4]
```diff
@@ -63,3 +69,52 @@ pub struct GltfPrimitive {
...
+#[derive(Default, Clone, TypeUuid, Debug)]
+#[uuid = "d81b7179-0448-4eb0-89fe-c067222725bf"]
+pub struct GltfAnimation {
+    pub node_animations: HashMap<usize, Vec<GltfNodeAnimation>>,
+}
+
+/// Key frames of an animation.
+#[derive(Clone, Debug)]
+pub enum GltfNodeAnimationKeyframes {
+    Rotation(Vec<Vec4>),
```
Comment (superdump) [inline crates/bevy_gltf/src/lib.rs:100]: Quat

File: crates/bevy_gltf/src/loader.rs [hunk 5]
```diff
@@ -109,6 +112,76 @@ async fn load_gltf<'a, 'b>(
...
+            };
+
+            let keyframes = if let Some(outputs) = reader.read_outputs() {
+                match outputs {
+                    gltf::animation::util::ReadOutputs::Translations(tr) => {
+                        GltfNodeAnimationKeyframes::Translation(tr.map(Vec3::from).collect())
+                    }
+                    gltf::animation::util::ReadOutputs::Rotations(rots) => {
+                        GltfNodeAnimationKeyframes::Rotation(
+                            rots.into_f32().map(Vec4::from).collect(),
```
Comment (superdump) [inline crates/bevy_gltf/src/loader.rs:149]: Quat

**Review by james7132 — APPROVED** [review]
LGTM. This should allow easy conversion into variable-rate animation curves rather easily. Main thing I'm a bit concerned with is the memory use of uncompressed values, but that can be addressed in a later PR.

**Review by cart — APPROVED** [review]
This looks good to me. I think its important to make it clear that the example is a "manual" implementation of an animation player and not a final bevy animation api, so I renamed it and added a doc comment explaining the situation.

**Pull request closed without merging**